### PR TITLE
Change: auth strategy WIP

### PR DESCRIPTION
**Pull request by MonietsSemen**

### CODE REVIEW

File: src/controllers/auth.controller.ts [hunk 1]
```diff
@@ -1,17 +1,30 @@
...
+    passport.authenticate(
+      'local',
+      { session: false },
+      async (err: Error, user: User, info: any) => {
+        if (err) return next(err);
+        if (!user) return res.redirect(`${env.userUrl}?error=invalid_credentials`);
+
+        const token = await AuthController.generateToken(user);
+        res.json({
+          your_private_token_for_next_time: {
```
Comment (DenisRupp) [inline src/controllers/auth.controller.ts]: delete `your_private_token_for_next_time`

File: src/controllers/order.controller.ts [hunk 2]
```diff
@@ -51,6 +65,14 @@ class OrderController {
 
     res.json({ orders });
   }
+
+  static getUser(req: Request, res: Response, _next: NextFunction) {
+    const localUser = (req.user as User)?.dataValues;
```
Comment (DenisRupp) [inline src/controllers/order.controller.ts]: don't use dataValues it's unnecessary

File: src/controllers/order.controller.ts [hunk 3]
```diff
@@ -51,6 +65,14 @@ class OrderController {
 
     res.json({ orders });
   }
+
+  static getUser(req: Request, res: Response, _next: NextFunction) {
```
Comment (DenisRupp) [inline src/controllers/order.controller.ts]: getUser in each controller? mb you can find better solution without repeating code each time.

File: src/controllers/user.controller.ts [hunk 4]
```diff
@@ -19,21 +19,26 @@ class UserController {
...
-      where: {
-        userId: user.id,
-      },
-    });
+    const newResponse: NewResponse = {
+      user: res.locals.user,
+    };
 
-    res.json({ user: res.locals.user, userOrders: orders });
+    if (localUser.role !== 'admin')
```
Comment (DenisRupp) [inline src/controllers/user.controller.ts]: Use middleware for this

File: src/controllers/user.controller.ts [hunk 5]
```diff
@@ -61,6 +66,14 @@ class UserController {
 
     res.status(NO_CONTENT).send();
   }
+
+  static getUser(req: Request, res: Response, _next: NextFunction) {
+    const localUser = (req.user as User)?.dataValues;
```
Comment (DenisRupp) [inline src/controllers/user.controller.ts]: why do you use optional chaining after declaring strict type?

**Review by DenisRupp — CHANGES_REQUESTED** [review]

File: src/controllers/user.controller.ts (hunk 5 above)
Comment (MonietsSemen) [reply src/controllers/user.controller.ts]: I was have some problems, but now it`s okay.

File: src/middlewares/authorization.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { NextFunction, Request, Response } from 'express';
+import { GetUser, SafeController } from '@/controllers/decorators';
+
+export default class Authorization {
+  @GetUser
+  @SafeController
+  static async roleVerification(_req: Request, res: Response, next: NextFunction) {
```
Comment (DenisRupp) [inline src/middlewares/authorization.ts:31]: Middleware should accept role as param.

File: src/controllers/decorators.ts [hunk 7]
```diff
@@ -36,3 +37,21 @@ export function CreatePagination(controllerMethod: any, context: ClassMethodDeco
...
   };
 }
+
+export function GetUser(controllerMethod: any, context: ClassMethodDecoratorContext) {
+  return async function wrapper(this: any, ...args: [Request, Response, NextFunction]) {
+    const [req, res, next] = args;
+    try {
+      const localUser = req.user;
+
+      if (!localUser) throw res.status(NOT_FOUND).send();
```
Comment (DenisRupp) [inline src/controllers/decorators.ts]: if not user it's unauthorized error (401)

**Review by DenisRupp — APPROVED** [review]

**Pull request merged**